File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution(object):
    def topKFrequent(self, nums, k):
        if len(nums) == 1:
            return nums
        #solve it by bucket sort
        freq = [[] for i in range(len(nums) + 1)]
        # print(freq)

        #create our {} to count the numbers
        counter = {}
        for n in nums:
            counter[n] = 1 + counter.get(n, 0)
        
        # start adding our counts for each number inside the freq depends on the index
        for number, count in counter.items():
            freq[count].append(number)
            # print(freq)
        
        result = [] #start from last i, stop at 0, decrement by 1
        for i in range(len(freq) -1, 0, -1):
            for n in freq[i]:
                result.append(n)
                if len(result) == k:
                    return result
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {

        int index = 0;
        int n = nums.length;

        HashMap<Integer, Integer> map = new HashMap<>();
        int[] buckets = new int[k];

        for (int number : nums) {

            map.put(number, map.getOrDefault(number, 0) + 1);
        }

        for (int i = n; i > 0; i--) {

            for (Map.Entry<Integer, Integer> entry : map.entrySet()) {

                int val = entry.getValue();
                int key = entry.getKey();

                if (val == i) {

                    if (index < k) {

                        buckets[index] = key;
                        index++;
                    } else {
                        break;
                    }
                }
            }

        }

        return buckets;
    }
}